transmart-api-server/src/main/groovy/org/transmartproject/api/server/user/KeycloakUserResourceService.groovy

```diff
@@ -0,0 +1,136 @@
+package org.transmartproject.api.server.user
+
+import com.google.common.collect.ImmutableMultimap
+import com.google.common.collect.Multimap
+import groovy.util.logging.Slf4j
+import org.springframework.beans.factory.annotation.Autowired
+import org.springframework.context.annotation.Primary
+import org.springframework.security.oauth2.provider.OAuth2Authentication
+import org.springframework.stereotype.Component
+import org.transmartproject.core.exceptions.NoSuchResourceException
+import org.transmartproject.core.users.AuthorisationChecks
+import org.transmartproject.core.users.ProtectedOperation
+import org.transmartproject.core.users.ProtectedResource
+import org.transmartproject.core.users.User
+import org.transmartproject.core.users.UsersResource
+
+import java.security.Principal
+import java.text.ParseException
+
+@Component
+@Primary
+@Slf4j
+class KeycloakUserResourceService implements UsersResource {
+
+    @Autowired
+    AuthorisationChecks authorisationChecks
+
+    @Override
+    User getUserFromUsername(String username) throws NoSuchResourceException {
+        throw new UnsupportedOperationException()
+    }
+
+    @Override
+    List<User> getUsers() {
+        throw new UnsupportedOperationException()
+    }
+
+    @Override
+    List<User> getUsersWithEmailSpecified() {
+        throw new UnsupportedOperationException()
+    }
+
+    @Override
+    User getUserFromPrincipal(Principal principal) {
+        assert principal instanceof OAuth2Authentication
+        def authentication = principal.userAuthentication
+        assert authentication: 'User is not authenticated.'
+
+        final String username = principal.name
+        List<String> authorities = principal.authorities*.authority
+        final boolean admin = authorities.remove('ROLE_ADMIN')
+        Multimap<String, ProtectedOperation> accessStudyTokenToOperations =
+                buildStudyTokenToOperationsMultimap(authorities, username)
+
+        def details = authentication.details
+        final String realName
+        final String email
+        if (details instanceof Map) {
+            realName = details.name
+            email = details.email
+        } else {
+            log.warn("Unexpected user details object for ${username} user. Expected map but was ${details}. Hence email and name can't be parsed.")
+            realName = null
+            email = null
+        }
+
+        createUser(username, realName, email, admin, accessStudyTokenToOperations)
+    }
+
+    private User createUser(final String username,
+                            final String realName,
+                            final String email,
+                            final boolean admin,
+                            final Multimap<String, ProtectedOperation> accessStudyTokenToOperations) {
+        new User() {
```

Comment:
I don't really like this anonymous class style. Why not just create a regular class for this?

Reply:
We probably should do this. How `canPerform(...)` method implementation would look like then?